llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -11393,6 +11422,53 @@ static SDValue combineMinNumMaxNumImpl(const SDLoc &DL, EVT VT, SDValue LHS,
   }
 }
 
+SDValue DAGCombiner::foldShiftToAvg(SDNode *N) {
+  const unsigned Opcode = N->getOpcode();
+
+  // Convert (sr[al] (add n[su]w x, y)) -> (avgfloor[su] x, y)
+  if (Opcode != ISD::SRA && Opcode != ISD::SRL)
```

Comment:
Do we need to check that the shift is by 1?

Reply:
The sd_match() below should catch it - it'd be nice to ensure we have negative test coverage for this though

Reply:
Indeed the `m_One()` from the pattern won't match.
Negative tests added as requested ;)

Reply:
Ah I see - I had missed that in the matcher and expected it to be up here.